### DISCUSSION TITLE
Alternate distance formulas

### DISCUSSION THREAD

**Discussion by markalefebvre** [moj-analytical-services/splink#208] [Q&A] [answered]
Is there a way to incorporate other distance formulas into the settings?  For instance, jaro winkler, levenshtein simliarity, cosine similarity?

**Comment by mamonu** [comment, marked as answer]
Hi. All the similarities you mentioned (jaro winkler+cosine similarity) are available if you have imported the appropriate jar and registered it. 

jar is here : https://github.com/moj-analytical-services/splink/blob/master/jars/scala-udf-similarity-0.0.7.jar


you can import the jar in this way in your sparkSession:

```python
spark = (
  SparkSession.builder.master('local[*]')
    .appName('blah')
    .config('spark.driver.extraClassPath', '/PATH/TO/scala-udf-similarity-0.0.7.jar')
    .config('spark.jars', '/PATH/TO/scala-udf-similarity-0.0.7.jar')
    .getOrCreate()
  )
```

a function to register all jar functions is  shown below:

```python
def add_udfs(spark):
    # Register UDFs
    udfs = [
        ("jaro_winkler_sim", "JaroWinklerSimilarity", DoubleType()),
        ("jaccard_sim", "JaccardSimilarity", DoubleType()),
        ("cosine_distance", "CosineDistance", DoubleType()),
 
        ("Dmetaphone", "DoubleMetaphone", StringType()),
        ("DmetaphoneAlt", "DoubleMetaphoneAlt", StringType()),
        ("QgramTokeniser", "QgramTokeniser", StringType()),
    ]

    for a, b, c in udfs:
        spark.udf.registerJavaFunction(a, "uk.gov.moj.dash.linkage." + b, c)

```

Then you can use them as you wish by creating a custom case expression on the settings
Levenstein is included in pyspark so its easy to include as well.


An example of a case statement incorporating levenstein 
(in the form of a function that creates sql code for the case statement) follows:

```python
def dob_case_statement_leven(dob_colname, leven_distance=1):
    # It's often the case that too many dates of birth are on the first of january
    # typically because when exact dob is unavailable they round to nearest year
    return f"""
    case
    when {dob_colname}_l is null or {dob_colname}_r is null then -1
    when {dob_colname}_l = {dob_colname}_r  and substr({dob_colname}_l, -5) = '01-01'  then 1
    when {dob_colname}_l = {dob_colname}_r  then 3
    when levenshtein({dob_colname}_l, {dob_colname}_r) <= {leven_distance} then 2
    else 0 end
    """
```

An example of a case expression using cosine sim:

```python

cosine_sim_expression = """

case when FULLNAME_l is null or FULLNAME_r is null then -1 
     when (1- cosine_distance(QgramTokeniser(FULLNAME_l), QgramTokeniser(FULLNAME_r))) > 0.84 then 2 
     when (1-cosine_distance(QgramTokeniser(FULLNAME_l), QgramTokeniser(FULLNAME_r))) > 0.70 then 1 
     else 0 end  as gamma_FULLNAME          

"""


```

**Reply by lenroc14** [reply]
Hello,

This is very helpful in the case that I would use the Jaro-Winkler similarity, Jaccard similarity, Cosine distance, QGramTokeniser, or some other string comparator that is defined and included in the .jar. However, if I wanted to define my own string comparator, and use this within Splink is there a way of doing this?

- Is there a way of writing a string comparator function, in python for instance, and applying this function within a custom case expression? Here is an arbitrary example of my objective: 

```
def CustomFunction(stringA, stringB):
     ------
     ------
```
```
custom_case_expression_with_custom_function = """

case when FULLNAME_l is null or FULLNAME_r is null then -1 
     when CustomFunction(FULLNAME_l, FULLNAME_r) > 0.70 then 1
     else 0    

"""
```


Thank you.

**Comment by mamonu** [comment]
There are some ways. 
Its a matter of speed on which is the best. 

- Is the string comparator function able to be expressed in SQL?  if yes then you just use it immediately in a macro-like function in the case expression
(an sql f-string with the parameters you want). An example ... you want the distance of two sets of lat-longs ,
so you create a sql function to calculate it. you then call it from your case expression

- Is it something that cannot be expressed in SQL? then the question is .... Do you have large datasets?
If the answer to both is yes then one efficient way is to program a scala jar that  performs this function 
Then import and register this function so its available from spark sql and then use it.

**Comment by mamonu** [comment]
@brittstenekes ,


To go a bit in more detail. 
The reason as you probably know that we have to use either Java or Scala to create the functions is that because the work in the executors has to be either in a JVM language or else a lot of context switching happens otherwise
![PySpark-Internals-Spark](https://user-images.githubusercontent.com/3820011/164520582-22fbfd4d-3f88-4faa-a868-b9235caf579b.png)


that is problematic time-wise for bigger amounts of data.

If you have a function that you think it cannot be done in SQL (and there are many of them!) I can propose two solutions.

- 1  ) if you think this is going to be a useful function to the wider community you can discuss about us including it in our 
[jar](https://github.com/moj-analytical-services/splink_scalaudfs).
If you have an algorithm working in another language such as Python/C++/ ??  we can discuss on how to port it in Scala or Java.

- 2 ) If you would like to create a function that cannot be shared for any reasons (for example because of privacy concerns) then very quickly i can propose some ways of how to create your own function that then is compiled into a jar and then registered through `spark.udf.registerJavaFunction` and in this way available through pysparksql so it can be used easily on a custom case expression

  - 2a) [here](https://stackoverflow.com/questions/51797395/spark-2-3-java-functions-callable-from-pyspark-python) is a Stack Overflow answer to a question that is kinda related. Once you create a Java function that does what you want you can use it in the way described there.

   - 2b)  [here](https://github.com/moj-analytical-services/splink_scalaudfs) you can see how we have done it in Scala.
Just copy the project somewhere and delete all our stuff and then write your own having a look on how types and functions are declared as an example.


Hope the above will be of help

**Reply by lenroc14** [reply]
Hello @mamonu,

Thank you for your detailed response. It is very informative. 

I have been successful at defining a Python UDF, registering the UDF through spark, and then using the UDF in a custom case expression.
Your notes on the Scala and Java functions are very useful as well and next steps will include converting these python UDFs to be used in Scala and Java. I will be looking into this in greater detail.

I have one remaining question:
When the UDF, CustomFunction, is called in a custom case expression as follows: 
```
custom_case_expression_with_custom_function = """

case when CustomFunction(FULLNAME_l, FULLNAME_r) is null then -1 
     when CustomFunction(FULLNAME_l, FULLNAME_r) > 0.70 then 1
     else 0    

"""
```

Is it possible to store the return value of CustomFunction? I see that Splink adds a 'gamma' column to indicate '-1','0', or '1' depending on the result of the case statement. Ideally, I would also have a column in the dataframe that indicates the return value of CustomFunction. 

Thank you for your assistance.

**Reply by RobinL** [reply]
Hiya.  If I undersatnd the question you're asking whether you could have custom case expressoin which just read `UDF()`, where `UDF` returns -1, 0 ,1 etc directly, rather than needing  `case when UDF() then` etc.
 

Sorry, you can't do this - this is a limitation of Splink.  it's necessary for various reasons - most notably ensuring error messages are sensible, and that we can plot various charts that come out of splink

We have thought about this quite a bit before though, because what you're trying to do makes sense! We just couldn't figure out how to integrate it into the rest of the codebase in a sensible way

**Comment by RobinL** [comment]
If the datasets you're linking are small, another option is to write is as a Python UDF:
https://sparkbyexamples.com/pyspark/pyspark-udf-user-defined-function/

This will work fine, it's just it will execute slowly.  But if your data is small, that might not be a problem
